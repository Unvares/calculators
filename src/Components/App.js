import React from 'react';
import { Slider } from './Slider';
import calculators from '../Assets/calculators.json';

const App = () => <Slider calculators={calculators} />;

export { App };
