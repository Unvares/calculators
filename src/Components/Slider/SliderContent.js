import React from 'react';
import { motion, AnimatePresence } from 'framer-motion';
import { Calculator } from '../Calculator';
import './SliderContent.scss';

const SliderContent = ({ activeId, onClick, calculator, length }) => {
  const prevActive = (activeId + length - 1) % length;
  const nextActive = (activeId + 1) % length;

  return (
    <AnimatePresence>
      <motion.div
        initial={{ scale: 0.75 }}
        animate={{ scale: 1 }}
        className='content'
      >
        <div
          className='content__arrow content__arrow_prev'
          onClick={() => onClick(prevActive)}
        >
          <span>&lt;</span>
        </div>
        <Calculator calculator={calculator} activeId={activeId} />
        <div
          className='content__arrow content__arrow_next'
          onClick={() => onClick(nextActive)}
        >
          <span>&gt;</span>
        </div>
      </motion.div>
    </AnimatePresence>
  );
};

export { SliderContent };
