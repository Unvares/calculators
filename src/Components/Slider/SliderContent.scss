.content {
  flex-grow: 1;
  display: flex;
  justify-content: center;
  align-items: center;
}

$arrow-size: 64px;

.content__arrow {
  border: 3px solid var(--additional-color);
  border-radius: 32px;
  cursor: pointer;
  font-size: 2rem;
  font-weight: bold;
  text-align: center;
  user-select: none;
  width: $arrow-size;
  height: $arrow-size;
  margin: 0 100px;
  transition: 0.3s ease;

  &:hover {
    border-color: var(--additional-color-2);
    color: var(--additional-color-2);
  }

  span {
    vertical-align: middle;
  }
}
