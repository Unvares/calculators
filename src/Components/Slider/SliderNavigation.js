import React from 'react';
import { useState, useRef, useEffect } from 'react';
import { motion } from 'framer-motion';
import cn from 'classnames';
import './SliderNavigation.scss';

const SliderNavigation = ({
  calculators,
  activeState: { activeId, setActive },
}) => {
  const [width, setWidth] = useState(0);
  const ref = useRef();

  useEffect(() => {
    const { offsetWidth, scrollWidth } = ref.current;
    setWidth(offsetWidth - scrollWidth);
  }, []);

  const updateActive = (index) => () => setActive(index);

  const names = calculators.map((calculator, index) => {
    const { name } = calculator;
    const elementClasses = cn({
      navigation__element: true,
      navigation__element_active: index === activeId,
    });
    return (
      <div className='navigation__wrapper' key={name}>
        <p className={elementClasses} onClick={updateActive(index)}>
          {name}
        </p>
      </div>
    );
  });

  return (
    <motion.div
      ref={ref}
      drag='x'
      dragConstraints={{
        left: width,
        right: 0,
      }}
      className='navigation'
    >
      {names}
    </motion.div>
  );
};

export default SliderNavigation;
