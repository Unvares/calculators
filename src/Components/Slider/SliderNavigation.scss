.navigation-wrapper::before {
  content: '';
  position: absolute;
  left: 0;
  top: var(--navigation-height);
  width: 100%;
  height: 1px;
  background-color: var(--additional-color);
}

.navigation {
  display: flex;
  align-content: stretch;
  flex-flow: row nowrap;
  height: var(--navigation-height);
  user-select: none;
}

.navigation__element-wrapper {
  display: flex;
  align-items: center;
  padding: 0 20px;
  flex-shrink: 0;
}

.navigation__element {
  cursor: pointer;
  color: transparent;
  background: linear-gradient(to bottom, white 50%, var(--additional-color-2) 50%);
  background-size: 100% 200%;
  background-clip: text;
  transition: background-position 250ms ease;
}

.navigation__element_active {
  background-position: 100% 100%;
  transition: background-position 250ms ease 100ms;
}
