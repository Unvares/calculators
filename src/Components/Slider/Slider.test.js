import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import { SliderContent } from './SliderContent';
import { SliderNavigation } from './SliderNavigation';
import { Slider } from '.';
import { Server } from '../App';

jest.mock('../App');

const calculators = [
  {
    title: 'Triangle Surface',
    inputs: ['height', 'base'],
    formula: '(H * B) / 2',
  },
  {
    title: 'Square Surface',
    inputs: ['side'],
    formula: 'A * A',
  },
  {
    title: 'Rectangle Surface',
    inputs: ['side 1', 'side 2'],
    formula: 'A * B',
  },
];

describe('SliderContent', () => {
  it('renders calculator and control arrows', () => {
    const {
      container: { firstChild: tree },
    } = render(
      <SliderContent
        activeId={0}
        calculator={calculators[0]}
        length={calculators.length}
      />
    );

    expect(tree).toMatchSnapshot();
  });
});

describe('SliderNavigation', () => {
  const titles = calculators.map(({ title }) => title);

  it('renders titles properly', () => {
    const {
      container: { firstChild: tree },
    } = render(<SliderNavigation titles={titles} />);

    expect(tree).toMatchSnapshot();
  });
});

describe('Slider', () => {
  const setup = () => {
    const {
      container: { firstChild: tree },
      findByDisplayValue,
      findByLabelText,
      findByText,
      getByText,
    } = render(<Slider calculators={calculators} />);

    return { tree, findByDisplayValue, findByLabelText, findByText, getByText };
  };

  it('returns null if no calculator data is provided', () => {
    const {
      container: { firstChild: tree },
    } = render(<Slider />);

    expect(tree).toBeNull();
  });

  it('switches via arrows properly', async () => {
    const { tree, findByText } = setup();

    const prevArrow = tree.querySelector('.content__arrow_prev');
    const nextArrow = tree.querySelector('.content__arrow_next');

    await findByText('TRIANGLE SURFACE');

    fireEvent.click(nextArrow);

    await findByText('SQUARE SURFACE');

    fireEvent.click(nextArrow);

    await findByText('RECTANGLE SURFACE');

    fireEvent.click(nextArrow);

    await findByText('TRIANGLE SURFACE');

    fireEvent.click(nextArrow);
    fireEvent.click(nextArrow);
    fireEvent.click(nextArrow);
    fireEvent.click(nextArrow);

    await findByText('SQUARE SURFACE');

    fireEvent.click(prevArrow);
    fireEvent.click(prevArrow);

    await findByText('RECTANGLE SURFACE');
  });

  it('switches via navigation panel properly', async () => {
    const { findByText, getByText } = setup();

    const switchToTriangleSurfaceButton = getByText('Triangle Surface');
    const switchToSquareSurfaceButton = getByText('Square Surface');
    const switchToRectangleSurfaceButton = getByText('Rectangle Surface');

    fireEvent.click(switchToSquareSurfaceButton);

    await findByText('SQUARE SURFACE');

    fireEvent.click(switchToRectangleSurfaceButton);

    await findByText('RECTANGLE SURFACE');

    fireEvent.click(switchToTriangleSurfaceButton);

    await findByText('Triangle Surface');
  });

  it('remembers last values properly', async () => {
    Server.fetch.mockResolvedValueOnce('25');
    Server.fetch.mockRejectedValueOnce('sus');

    const { tree, findByDisplayValue, findByLabelText, findByText } = setup();
    const nextArrow = tree.querySelector('.content__arrow_next');

    const inputBase = await findByLabelText('BASE');
    const inputHeight = await findByLabelText('HEIGHT');
    const submitButton = tree.querySelector('button[type="submit"]');

    fireEvent.change(inputBase, { target: { value: 10 } });
    fireEvent.change(inputHeight, { target: { value: 5 } });
    fireEvent.click(submitButton);

    await findByText('25');

    fireEvent.click(nextArrow);

    const inputSide = await findByLabelText('SIDE');

    fireEvent.change(inputSide, { target: { value: 'Not a number' } });
    fireEvent.click(submitButton);

    await findByText('ERROR');

    fireEvent.click(nextArrow);
    fireEvent.click(nextArrow);

    await findByText('25');
    await findByDisplayValue('10');
    await findByDisplayValue('5');

    fireEvent.click(nextArrow);

    await findByDisplayValue('Not a number');
    await findByText('ERROR');
  });
});
