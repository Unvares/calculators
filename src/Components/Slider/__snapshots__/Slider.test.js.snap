// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SliderContent renders calculator and control arrows 1`] = `
<div
  class="content"
  draggable="false"
  style="transform: scale(0.75) translateZ(0); user-select: none;"
>
  <div
    class="content__arrow content__arrow_prev"
  >
    <div
      class="content__arow_top"
    />
    <div
      class="content__arow_bottom"
    />
  </div>
  <div
    class="calculator"
  >
    <h2
      class="calculator__text calculator__text_engraved"
    >
      TRIANGLE SURFACE
    </h2>
    <form>
      <div
        class="calculator__line calculator__line_input"
      >
        <label
          class="calculator__text calculator__text_engraved"
          for="height"
        >
          HEIGHT
        </label>
        <input
          autocomplete="off"
          class="calculator__input calculator__text"
          id="height"
          name="height"
          type="text"
          value=""
        />
      </div>
      <div
        class="calculator__line calculator__line_input"
      >
        <label
          class="calculator__text calculator__text_engraved"
          for="base"
        >
          BASE
        </label>
        <input
          autocomplete="off"
          class="calculator__input calculator__text"
          id="base"
          name="base"
          type="text"
          value=""
        />
      </div>
      <button
        style="display: none;"
        type="submit"
      />
    </form>
    <section
      class="calculator__info"
    >
      <div
        class="calculator__line"
      >
        <span
          class="calculator__text calculator__text_engraved"
        >
          FORMULA
        </span>
        <span
          class="calculator__text calculator__text_engraved calculator__text_input-aligned"
        >
          (H * B) / 2
        </span>
      </div>
      <div
        class="calculator__line"
      >
        <span
          class="calculator__text calculator__text_engraved"
        >
          RESULT
        </span>
        <span
          class="calculator__text calculator__text_input-aligned"
        />
      </div>
    </section>
  </div>
  <div
    class="content__arrow content__arrow_next"
  >
    <div
      class="content__arow_top"
    />
    <div
      class="content__arow_bottom"
    />
  </div>
</div>
`;

exports[`SliderNavigation renders titles properly 1`] = `
<div
  class="navigation-wrapper"
  data-projection-id="1"
  style="transform: translate(-100px);"
>
  <div
    class="navigation"
    data-projection-id="2"
    draggable="false"
    style="user-select: none;"
  >
    <div
      class="navigation__element-wrapper"
    >
      <p
        class="navigation__element"
      >
        Triangle Surface
      </p>
    </div>
    <div
      class="navigation__element-wrapper"
    >
      <p
        class="navigation__element"
      >
        Square Surface
      </p>
    </div>
    <div
      class="navigation__element-wrapper"
    >
      <p
        class="navigation__element"
      >
        Rectangle Surface
      </p>
    </div>
  </div>
</div>
`;
